src/setup/stateful.rs

```diff
@@ -0,0 +1,72 @@
+//! Contains code to start and stop a node with preloaded ledger data.
+
+use std::{
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4},
+    path::PathBuf,
+    process::Child,
+};
+
+use crate::setup::{config::NodeMetaData, process, process::wait_for_start};
+
+const STATEFUL_PORT: u16 = 5005;
+
+/// Stateful node. Binds to 127.0.0.1:5005, loads data and configuration from `~/.ziggurat/ripple_stateful`.
+/// Does *not* clean its database after stopping.
+pub struct StatefulNode {
+    /// The process encapsulating the running node.
+    process: Option<Child>,
+}
+
+impl StatefulNode {
+    /// Build and starts a new stateful node.
+    pub async fn new(config_path: PathBuf, log_to_stdout: bool) -> anyhow::Result<Self> {
+        let mut meta = NodeMetaData::new(config_path)?;
+        meta.start_args.append(&mut vec![
+            "--valid".into(),
+            "--quorum".into(),
+            "1".into(),
+            "--load".into(),
+        ]);
+        let process = process::start(&meta, log_to_stdout);
+        wait_for_start(&SocketAddr::V4(SocketAddrV4::new(
+            Ipv4Addr::LOCALHOST,
+            STATEFUL_PORT,
+        )))
+        .await;
+        Ok(Self {
+            process: Some(process),
+        })
+    }
+
+    /// Stops the node. This *does not* clean node's cache.
+    pub fn stop(&mut self) {
+        if let Some(message) = self.process.take().and_then(process::stop) {
+            eprintln!("Child exited with message: {}", message)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{io, path::PathBuf, time::Duration};
+
+    use crate::setup::{config::ZIGGURAT_DIR, stateful::StatefulNode};
+    const STATEFUL_DIR: &str = "ripple_stateful";
+
+    #[tokio::test]
+    #[ignore = "convenience test to tinker with a running node for dev purposes"]
+    async fn should_start_stop_stateful_node() {
+        let path = build_stateful_path().unwrap();
+        let mut node = StatefulNode::new(path, true).await.unwrap();
+        tokio::time::sleep(Duration::from_secs(10)).await;
+        node.stop();
+    }
+
+    // Convenience function to build stateful node's path.
+    fn build_stateful_path() -> io::Result<PathBuf> {
+        Ok(home::home_dir()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "couldn't find home directory"))?
+            .join(ZIGGURAT_DIR)
+            .join(STATEFUL_DIR))
```

Comment:
we should instead copy the contents of this dir to some temporary directory, to remove the possibility of polluting the saved test state; that way the stateful nodes can still remove their _own_ state after every test

Reply:
Done